### PR TITLE
Use type-directed lowering

### PR DESCRIPTION
**Pull request by zardyh**
Instead of guessing when things should be quantified, we can use the
fact that we statically know the types of everything to figure out
precisely when things should be quantified.

This does that.

### CODE REVIEW

File: src/Core/Lower.hs [hunk 1]
```diff
@@ -50,101 +48,78 @@ errRef = CotRef (TgInternal "error")
...
+lowerAnyway :: MonadLower m => Expr Typed -> m CoTerm
+lowerAnyway (Record xs _) = case xs of
+  [] -> pure (CotLit ColRecNil)
+  xs -> do
+    xs' <- for xs $ \(label, ex) ->
+      (,,) <$> pure label <*> lowerType (getType ex) <*> lowerExpr ex
+    pure (CotExtend (CotLit ColRecNil) xs')
+lowerAnyway (RecordExt e xs _) = do
+  xs' <- for xs $ \(label, ex) ->
+    (,,) <$> pure label <*> lowerType (getType ex) <*> lowerExpr ex
```
Comment (SquidDev) [inline src/Core/Lower.hs]: Tuple sections? I know you don't like them, so fair doos if not.

Comment (zardyh) [reply src/Core/Lower.hs]: Eh, sure.

**Comment by zardyh** [comment]
I need to see if this lets me remove my _favourite_ hack from the compiler, [this](https://github.com/zardyh/amulet/pull/6#discussion_r168627743)

**Comment by zardyh** [comment]
It does!

**Comment by zardyh** [comment]
This is done.

**Pull request merged**